core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,402 @@
+      Configuration.getBoolean(PropertyKey.USER_NETWORK_NETTY_READER_CANCEL_ENABLED);
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_READER_BUFFER_SIZE_PACKETS);
+  private static final long READ_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final Channel mChannel;
+  private final Protocol.RequestType mRequestType;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mStart;
+  private final long mBytesToRead;
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  @GuardedBy("mLock")
+  private final Queue<ByteBuf> mPackets = new LinkedList<>();
+  @GuardedBy("mLock")
+  private Throwable mPacketReaderException = null;
+  private final Condition mNotEmptyOrFail = mLock.newCondition();
+
+  /** The next pos to read. */
+  private long mPosToRead;
+  /** This is true only when an empty packet is received. */
+  private boolean mDone = false;
+
+  private boolean mClosed = false;
+
+  /**
+   * Creates an instance of {@link NettyPacketReader}. If this is used to read a block remotely, it
+   * requires the block to be locked beforehand and the lock ID is passed to this class.
+   *
+   * @param address the netty data server network address
+   * @param id the block ID or UFS file ID
+   * @param offset the offset
+   * @param len the length to read
+   * @param lockId the lock ID
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException if it fails to create the object
+   */
+  public NettyPacketReader(InetSocketAddress address, long id, long offset, long len,
+      long lockId, long sessionId, Protocol.RequestType type) throws IOException {
+    Preconditions.checkArgument(offset >= 0 && len > 0);
+
+    mAddress = address;
+    mId = id;
+    mStart = offset;
+    mPosToRead = offset;
+    mBytesToRead = len;
+    mRequestType = type;
+
+    mChannel = BlockStoreContext.acquireNettyChannel(address);
+    addHandler();
+    Protocol.ReadRequest readRequest =
+        Protocol.ReadRequest.newBuilder().setId(id).setOffset(offset).setLength(len)
+            .setLockId(lockId).setSessionId(sessionId).setType(type).build();
+    mChannel.writeAndFlush(new RPCProtoMessage(readRequest))
+        .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  @Override
+  public ByteBuf readPacket() throws IOException {
+    Preconditions.checkState(!mClosed, "PacketReader is closed while reading packets.");
+    ByteBuf buf = null;
+    mLock.lock();
+    try {
+      while (true) {
+        if (mDone) {
+          return null;
+        }
+        if (mPacketReaderException != null) {
+          throw new IOException(mPacketReaderException);
+        }
+        buf = mPackets.poll();
+        if (!tooManyPacketsPending()) {
```

Comment:
Should there be a bit more lag for the resuming? Instead of resuming and pausing constantly, if there was more lag, there could be more amortization of state changes? This could be added as a TODO.